source/ngf-mtl/impl.cpp

```diff
@@ -448,6 +465,64 @@ static uint32_t ngfmtl_get_num_rows(const uint32_t height, const ngf_image_forma
 
 #pragma mark ngf_struct_definitions
 
+enum ngf_id_init_types { id_default };
+
+// Shared pointer for managed objects
+template<typename T> class ngf_id {
+  public:
+  // Create an ngf_id with an additional retain count
+  // Useful for keeping AutoReleasePool managed objects alive beyond
+  // Their pool lifetime
+  static ngf_id add_retain(T* ptr) {
+    ngf_id res = ptr;
+    if (res) { res->retain(); }
+    return res;
+  }
+
+  ngf_id() : ptr_(nullptr) {
+  }
+  ngf_id(const ngf_id_init_types& type) : ptr_(T::alloc()->init()) {
+  }
+  // Note: Does NOT increment ref count. You can use this directly after calling alloc()->init()
+  ngf_id(T* starting_ptr) : ptr_(starting_ptr) {
+  }
+  ~ngf_id() {
+    if (ptr_) { ptr_->release(); }
```

Comment:
i usually separate this de-initialization code into a `destroy_if_necessary` method and call it from both the dtor and copy/move assignment operators.

Reply:
Done